pkg/trace/info/stats.go

```diff
@@ -155,15 +167,108 @@ func (ts *TagStats) publish() {
 	metrics.Count("datadog.trace_agent.receiver.events_extracted", eventsExtracted, tags, 1)
 	metrics.Count("datadog.trace_agent.receiver.events_sampled", eventsSampled, tags, 1)
 	metrics.Count("datadog.trace_agent.receiver.payload_accepted", requestsMade, tags, 1)
+
+	for reason, count := range ts.TracesDropped.tagValues() {
+		metrics.Count("datadog.trace_agent.normalizer.traces_dropped", count, append(tags, "reason:"+reason), 1)
+	}
+
+	for reason, count := range ts.TracesMalformed.tagValues() {
+		metrics.Count("datadog.trace_agent.normalizer.traces_malformed", count, append(tags, "reason:"+reason), 1)
+	}
+}
+
+// inlineNonZeroTagValues serializes the entries in this map into format "key1: value1, key2: value2, ...", sorted by
+// key to ensure consistent output order. Only non-zero values are included.
+func inlineNonZeroTagValues(statsMap map[string]int64) string {
+	keys := make([]string, len(statsMap))
+	i := 0
+	for k := range statsMap {
+		keys[i] = k
+		i++
+	}
+	sort.Strings(keys)
+
+	var results []string
+	for _, key := range keys {
+		value := statsMap[key]
+		if value > 0 {
+			results = append(results, fmt.Sprintf("%s:%d", key, value))
+		}
+	}
+
+	return strings.Join(results, ", ")
+}
+
+// TracesDropped contains counts for reasons traces have been dropped
+type TracesDropped struct {
+	DecodingError int64
+	EmptyTrace    int64
+	TraceIDZero   int64
+	SpanIDZero    int64
+	ForeignSpan   int64
+}
+
+func (s *TracesDropped) tagValues() (result map[string]int64) {
+	return map[string]int64{
+		"decoding_error": atomic.LoadInt64(&s.DecodingError),
+		"empty_trace":    atomic.LoadInt64(&s.EmptyTrace),
+		"trace_id_zero":  atomic.LoadInt64(&s.TraceIDZero),
+		"span_id_zero":   atomic.LoadInt64(&s.SpanIDZero),
+		"foreign_span":   atomic.LoadInt64(&s.ForeignSpan),
+	}
+}
+
+func (s *TracesDropped) String() string {
+	return inlineNonZeroTagValues(s.tagValues())
+}
+
+// TracesMalformed contains counts for reasons malformed traces have been accepted after applying automatic fixes
+type TracesMalformed struct {
+	DuplicateSpanID       int64
```

Comment:
Maybe we should document these, but I don't feel strongly about it. Just a thought...

Reply:
Yeah probably good to document. Added docs.